### PR TITLE
Add support for anywhere rules

### PR DESCRIPTION
**Pull request by yilongli**

### CODE REVIEW

File: src/javasources/KTool/src/org/kframework/backend/java/symbolic/SubstituteAndEvaluateTransformer.java [hunk 1]
```diff
@@ -25,6 +25,8 @@
  */
 public class SubstituteAndEvaluateTransformer extends CopyOnWriteTransformer {
 
+    private static final boolean ENABLE_DEBUG = false;
```
Comment (andreistefanescu) [inline src/javasources/KTool/src/org/kframework/backend/java/symbolic/SubstituteAndEvaluateTransformer.java]: we should use either -v or introduce a --debug flag

**Comment by yilongli** [comment]
I am sure not sure why the CIL semantics fails on type checking, but, anyway, it is ready for review. @andreistefanescu @dwightguth 

@grosu I changed the definitions of SIMPLE and KOOL in the tutorial to use [anywhere] rules as well as some comments. Please take a look.

**Comment by andreistefanescu** [comment]
I merged a cil commit that only works with Radu's branch, so now it is
broken.

File: src/javasources/KTool/src/org/kframework/backend/java/kil/Definition.java [hunk 2]
```diff
@@ -96,6 +98,15 @@ public void addRule(Rule rule) {
             }
         } else if (rule.containsAttribute(Attribute.MACRO_KEY)) {
             macros.add(rule);
+        } else if (rule.containsAttribute(Attribute.ANYWHERE_KEY)) {
+            if (!(rule.leftHandSide() instanceof KItem)) {
+                GlobalSettings.kem.registerCriticalWarning(
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/Definition.java:107]: Can you make an overload of the registerCriticalWarning method that takes only the non-null arguments here?

File: src/javasources/KTool/src/org/kframework/backend/java/kil/Rule.java [hunk 3]
```diff
@@ -175,13 +177,26 @@ public void visit(Cell cell) {
...
+            if (sortPredArg == null) {
+                isSortPredicate = false;
+
+                /*
+                 * YilongL: the right-hand side of the sort predicate rule
+                 * is not necessarily {@code BoolToken#True}? for example:
+                 *     rule isNat(I:Int) => '_>=Int_(I:Int,, Int(#"0"))
+                 */
+                // TODO(YilongL): properly re-implement support for sort predicate rules
+//                GlobalSettings.kem.registerCriticalWarning("Unexpected sort predicate rule: " + this, null, this);
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/Rule.java:198]: This todo worries me. Can you tell me what you mean by it? have you disabled some kind of functionality that was working before?

**Comment by dwightguth** [comment]
Lgtm other than the things I mentioned. @grosu can you review?

**Comment by dwightguth** [comment]
Jenkins: test this please

File: src/javasources/KTool/src/org/kframework/backend/java/kil/Rule.java (hunk 3 above)
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/Rule.java:198]: No, everything that worked before still works now. I am just excluding some new cases which I couldn't handle. Specifically, I used to assume that sort predicate rule is defined on KItem only, i.e., `rule isVal(cons V:Val) => true` in `fun-untyped.k`. However, now I see `rule isNat(I:Int) => I >=Int 0 [function, anywhere]` in `builtins/int.k` after including all `[anywhere]` rules. I need to exclude this rule until I properly re-implement the support for sort predicate later.

File: src/javasources/KTool/src/org/kframework/backend/java/kil/Definition.java (hunk 2 above)
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/Definition.java:107]: I have addressed this comment, but the change is actually on the next line. So the comment doesn't disappear.

**Comment by yilongli** [comment]
@dwightguth the build time increases to 30mins again. Is it because our recent changes to the k framework or the Jenkins issue you mentioned last time?

@grosu please review the changes to tutorial

**Comment by dwightguth** [comment]
The build seems to have spiked in time by about 3 minutes between builds
http://office.runtimeverification.com:8080/rv-jenkins/job/k-framework-pull-requests/224/
and
http://office.runtimeverification.com:8080/rv-jenkins/job/k-framework-pull-requests/225/

The logical conclusion is that either something changed on that box between
those two builds, or the change I made to machine integers is a performance
problem. I don't really understand how either of those things could be
true though... I will investigate on Monday.

File: tutorial/2_languages/1_simple/1_untyped/simple-untyped.k [hunk 4]
```diff
@@ -566,17 +566,15 @@ precise location; we prefer to do it this way to avoid locking the store.
...
-// would be underused, because it would only be used at the top of the
-// computation or inside the lvalue wrapper.  So it may not be worth,
-// or we may need to come up with a special notation allowing us to
-// enumerate contexts for [anywhere] rules.
-  rule V:Val[N1:Int, N2:Int, Vs:Vals] => V[N1][N2, Vs]            [structural]
-  rule lvalue(V:Val[N1:Int, N2:Int, Vs:Vals] => V[N1][N2, Vs])    [structural]
-
-  rule array(L,_)[N:Int] => lookup(L +Int N)            [structural]
-  rule lvalue(array(L,_)[N:Int] => lookup(L +Int N))    [structural]
+// The [anywhere] feature would be underused, because it would only
```
Comment (grosu) [inline tutorial/2_languages/1_simple/1_untyped/simple-untyped.k]: Here I would use present tense: "The [anywhere] feature is underused ...".

Other than this, all the changes to the tutorial look great.  I am very happy to see the anywhere rules back.  Thanks, Yilong!

**Pull request merged**